src/drivers/uavcan_v1/Subscribers/BaseSubscriber.hpp

```diff
@@ -53,28 +53,78 @@ class UavcanBaseSubscriber
 	static constexpr uint16_t CANARD_PORT_ID_UNSET = 65535U;
 
 	UavcanBaseSubscriber(CanardInstance &ins, const char *subject_name, uint8_t instance = 0) :
-		_canard_instance(ins), _subject_name(subject_name), _instance(instance) { };
+		_canard_instance(ins), _instance(instance)
+	{
+		_subj_sub._subject_name = subject_name;
+	}
 
 	virtual void subscribe() = 0;
-	virtual void unsubscribe() { canardRxUnsubscribe(&_canard_instance, CanardTransferKindMessage, _port_id); };
+	virtual void unsubscribe()
+	{
+		SubjectSubscription *curSubj = &_subj_sub;
+
+		while (curSubj != NULL) {
+			canardRxUnsubscribe(&_canard_instance, CanardTransferKindMessage, curSubj->_canard_sub._port_id);
+			curSubj = curSubj->next;
+		}
+	};
 
 	virtual void callback(const CanardTransfer &msg) = 0;
 
-	CanardPortID id() { return _port_id; };
+	CanardPortID id(uint32_t id = 0)
+	{
+		uint32_t i = 0;
+		SubjectSubscription *curSubj = &_subj_sub;
+
+		while (curSubj != NULL) {
+			if (id == i) {
+				return curSubj->_canard_sub._port_id;
+			}
```

Comment:
Would it help you if we added a `void* user_reference` to [`CanardRxSubscription`](https://github.com/UAVCAN/libcanard/blob/55938c51cf7da3096679933302693c9dae7a2b33/libcanard/canard.h#L272-L294)?

Not sure if a user of this API would expect the non-trivial variable complexity of these calls. I would worry that one might inadvertently call one of these methods from a tight loop.

Reply:
Actually I rather hide the CanardRxSubscription pointer from users, this function just returns the portID used by the class but you can have multiple instances, the parameter name id is confusing though I will change that.

Reply:
> Actually I rather hide the CanardRxSubscription pointer from users

I get it. What I'm asking is: if we added a [user reference](https://github.com/UAVCAN/libcanard/blob/55938c51cf7da3096679933302693c9dae7a2b33/libcanard/canard.h#L320), would you be able to leverage it to make your implementation simpler?